scipy/_lib/_util.py

```diff
@@ -140,3 +141,36 @@ def _asarray_validated(a, check_finite=True,
             raise ValueError('object arrays are not supported')
     return a
 
+
+def inspect_args(func):
+    """
+    Args/defaults from inspect.getargspec (py2k) or inspect.signature (py3k).
+
+    `inspect.signature` does not exist in Python 2.x, and `inspect.getargspec`
+    is deprecated in Python 3.x.  This function is a simple wrapper around both
+    that returns only the parameters and defaults lists (typically all that is
+    needed within Scipy).
+
+    Parameters
+    ----------
+    func : callable
+        The function to inspect.
+
+    Returns
+    -------
+    args : sequence of str
+        The argument names.  For Python 2.x, a list of strings.  For Python
+        3.x, an OrderedDict.
+    defaults : list
+        A list of defaults for keyword arguments.
+
+    """
+    if sys.version_info[0] >= 3:
+        sig = inspect.signature(func)
+        args = sig.parameters
+        tmp_defaults = [d.default for d in sig.parameters.values()]
+        defaults = [d for d in tmp_defaults if d is not inspect._empty]
+    else:
+        args, _, _, defaults = inspect.getargspec(func)
+
+    return args, defaults
```

Comment:
It should be marginally faster to conditionally define the function rather than check for the Python version in every call.